File: frontend/src/lib/eltm-transfer.ts
import type { EltmExportEntity, EltmExportNote, EltmExportPayload, EltmExportRelationship } from './types'

/**
 * Parse an exported ELTM file into the import payload with a minimal shape
 * check: the server owns the deep validation (dates parse, endpoints
 * resolve, values are single lines — see `EltmTransferService.importEltm`),
 * this only rejects obviously-wrong files before a request is spent, with
 * the file name in every error message.
 */
export async function parseEltmImportFile(file: File): Promise<EltmExportPayload> {
  let parsed: unknown
  try {
    parsed = JSON.parse(await file.text())
  } catch {
    throw new Error(`"${file.name}" is not valid JSON`)
  }
  const candidate = parsed as Partial<EltmExportPayload> | null
  if (
    !candidate ||
    typeof candidate.entities !== 'object' ||
    candidate.entities === null ||
    Array.isArray(candidate.entities) ||
    !Array.isArray(candidate.relationships)
  ) {
    throw new Error(
      `"${file.name}" is not an exported ELTM file (expected {"entities": {...}, "relationships": [...]})`,
    )
  }
  for (const [uuid, entity] of Object.entries(candidate.entities as Record<string, unknown>)) {
    if (!isEltmExportEntity(entity)) {
      throw new Error(
        `"${file.name}" is not an exported ELTM file (entity "${uuid}" must carry name, category, attributes and notes)`,
      )
    }
  }
  for (const relationship of candidate.relationships as unknown[]) {
    if (!isEltmExportRelationship(relationship)) {
      throw new Error(
        `"${file.name}" is not an exported ELTM file (each relationship must carry srcUuid, verb, dstUuid, valid and notes)`,
      )
    }
  }
  return candidate as EltmExportPayload
}

function isEltmExportEntity(candidate: unknown): candidate is EltmExportEntity {
  const entity = candidate as Partial<EltmExportEntity> | null
  return (
    !!entity &&
    typeof entity.name === 'string' &&
    typeof entity.category === 'string' &&
    isStringRecord(entity.attributes) &&
    Array.isArray(entity.notes) &&
    entity.notes.every(isEltmExportNote)
  )
}

function isEltmExportRelationship(candidate: unknown): candidate is EltmExportRelationship {
  const relationship = candidate as Partial<EltmExportRelationship> | null
  return (
    !!relationship &&
    typeof relationship.srcUuid === 'string' &&
    typeof relationship.verb === 'string' &&
    typeof relationship.dstUuid === 'string' &&
    typeof relationship.valid === 'boolean' &&
    Array.isArray(relationship.notes) &&
    relationship.notes.every(isEltmExportNote)
  )
}

function isEltmExportNote(candidate: unknown): candidate is EltmExportNote {
  const note = candidate as Partial<EltmExportNote> | null
  return !!note && typeof note.date === 'string' && typeof note.note === 'string'
}

function isStringRecord(candidate: unknown): candidate is Record<string, string> {
  if (typeof candidate !== 'object' || candidate === null || Array.isArray(candidate)) return false
  return Object.values(candidate).every((value) => typeof value === 'string')
}

File: src/main/kotlin/info/skyblond/daapu/memory/eltm/EltmTransfer.kt
package info.skyblond.daapu.memory.eltm

import kotlinx.serialization.Serializable

/**
 * One diary note in the transfer payload: the LLM-resolved absolute event
 * date (`YYYY-MM-DD`) plus the note text. No subject: the note nests
 * inside its entity/relationship entry, so no note ids exist at all.
 */
@Serializable
data class EltmExportNote(
    val date: String,
    val note: String,
)

/**
 * One entity entry of the transfer payload. [name] and [category] are the
 * stored canonical forms (already normalized — see [normalizeName]);
 * import matches on `(name, category)`, never on the payload's uuid.
 */
@Serializable
data class EltmExportEntity(
    val name: String,
    val category: String,
    /** Current-state facts, as stored (normalized keys, single-line values). */
    val attributes: Map<String, String>,
    val notes: List<EltmExportNote>,
)

/**
 * One relationship entry of the transfer payload. [srcUuid]/[dstUuid]
 * reference the payload's entity keys; [verb] is the stored canonical form
 * (see [normalizeVerb]); [valid] is the row's structural state AT EXPORT
 * TIME — import applies it only when the file's newest note is newer than
 * the target row's (see `EltmTransferService.importEltm`).
 */
@Serializable
data class EltmExportRelationship(
    val srcUuid: String,
    val verb: String,
    val dstUuid: String,
    val valid: Boolean,
    val notes: List<EltmExportNote>,
)

/**
 * The ELTM transfer payload: the shape of `GET /api/eltm/export`'s
 * attachment (`eltm.json`) AND the request body of `POST /api/eltm/import`
 * (the `overwriteAttr` decision travels as a query parameter, so the
 * exported file posts verbatim as the body — see `EltmRoute.kt`).
 *
 * Entities are keyed by a file-level uuid: minted fresh at every export, a
 * pure join key for the relationships' [EltmExportRelationship.srcUuid] /
 * [EltmExportRelationship.dstUuid] references — import NEVER matches on it
 * (entities match on `(name, category)`, relationships on the resolved
 * entity ids plus verb). No db ids and no embeddings anywhere: the import
 * re-derives everything (fresh row ids, embeddings recomputed through the
 * local hand), so a file transfers across instances with different
 * embedding models.
 */
@Serializable
data class EltmExportPayload(
    val entities: Map<String, EltmExportEntity>,
    val relationships: List<EltmExportRelationship>,
)

/**
 * The `POST /api/eltm/import` response: the merge's per-kind split.
 * [entitiesCreated]/[relationshipsCreated] count file entries whose row the
 * pre-import snapshot did NOT hold (normally this import's insert; in a
 * write race, a concurrent writer's row the create-or-fetch adopted);
 * [entitiesMatched]/[relationshipsMatched] count rows the snapshot already
 * held (see `EltmTransferService`'s concurrency stance).
 * [attributesWritten] counts attribute keys the import SET (a new key, or
 * an overwrite under `overwriteAttr=true` that changed the value);
 * [attributesKept] counts existing keys left alone (skipped under
 * `overwriteAttr=false`, or an identical value — the service's no-op).
 * [notesSkipped] counts exact (event date, trimmed text) duplicates, in the
 * database or earlier in the same file; [notesInserted] the rest.
 */
@Serializable
data class EltmImportSummary(
    val entitiesCreated: Int,
    val entitiesMatched: Int,
    val relationshipsCreated: Int,
    val relationshipsMatched: Int,
    val notesInserted: Int,
    val notesSkipped: Int,
    val attributesWritten: Int,
    val attributesKept: Int,
)
